addons/website_sale/controllers/main.py

```diff
@@ -1215,20 +1226,40 @@ def process_express_checkout(self, billing_address, **kwargs):
             child_partner_id = self._find_child_partner(
                 order_sudo.partner_id.commercial_partner_id.id, billing_address
             )
-            if child_partner_id:
-                order_sudo.partner_invoice_id = child_partner_id
-            else:
-                billing_partner_id = self._create_or_edit_partner(
-                    billing_address,
-                    type='invoice',
-                    parent_id=order_sudo.partner_id.id,
+            order_sudo.partner_invoice_id = child_partner_id or self._create_or_edit_partner(
+                billing_address, type='invoice', parent_id=order_sudo.partner_id.id
+            )
+
+        if shipping_address:
+
+            self._include_country_and_state_in_address(shipping_address)
+
+            if order_sudo.partner_shipping_id.name.endswith(order_sudo.name):
+                # The existing partner was created by `process_express_checkout_delivery_choice`, it
+                # means that the partner is missing information, so we update it.
+                order_sudo.partner_shipping_id = self._create_or_edit_partner(
+                    shipping_address,
+                    edit=True,
+                    type='delivery',
+                    partner_id=order_sudo.partner_shipping_id.id,
+                )
+            elif any(
+                shipping_address[k] != order_sudo.partner_shipping_id[k] for k in shipping_address
+            ):
+                # The sale order's shipping partner's address is different from the one received. If
+                # all the sale order's child partners' address differs from the one received, we
+                # create a new partner. The phone isn't always checked because it isn't sent in
+                # shipping information with Google Pay.
+                child_partner_id = self._find_child_partner(
+                    order_sudo.partner_id.commercial_partner_id.id, shipping_address
+                )
+                order_sudo.partner_shipping_id = child_partner_id or self._create_or_edit_partner(
+                    shipping_address, type='delivery', parent_id=order_sudo.partner_id.id
                 )
-                order_sudo.partner_invoice_id = billing_partner_id
 
-        # In a non-express flow, `sale_last_order_id` would be added in the session before the
-        # payment. As we skip all the steps with the express checkout, `sale_last_order_id` must be
-        # assigned to ensure the right behavior from `shop_payment_confirmation()`.
-        request.session['sale_last_order_id'] = order_sudo.id
+        # Process the delivery carrier
+        if shipping_option:
+            order_sudo._check_carrier_quotation(force_carrier_id=int(shipping_option['id']))
```

Comment:
Could it be possible to have a `shipping_adress` and no `shipping_option` in an obscure flow? Or the opposite?
If yes, then you introduced a comportment change, and I don't think it's intended.

Reply:
If you are sure _(and @chevalierv  confirms it as he knows these flows better)_, then fine be me, even if I would prefer to keep the `if shipping_address and shipping_option:` from before for now to be sure, until we rework that :shrug:

Reply:
Hello all :wave: 

First, it should be impossible to have no `shipping_option`. It will be in the PR @vava-odoo mentioned though. The opposite isn't technically possible with our current implementation.

I see no harm in this diff.